kcp/test.sh

```diff
@@ -0,0 +1,230 @@
+        ENVIRONMENT must be in [$(ls -F "$SCRIPT_DIR/environments" | grep "/$" | tr '/\n' ', ')local]
+        Default value: local
+    -k, --kcp KCP_CONTEXT
+        Select the KCP instance on which to test. Must be one of [stable, unstable].
+        Default value: unstable
+    -w, --workspace KCP_WORKSPACE
+        Select the KCP workspace to log into.
+        Default value: pln-svc
+    -d, --debug
+        Activate tracing/debug mode.
+    -h, --help
+        Display this message.
+
+Example:
+    ${0##*/} --all
+" >&2
+}
+
+parse_args() {
+  DEBUG=" "
+  ENVIRONMENT="local"
+  KCP_ENV="kcp-unstable"
+  KCP_WORKSPACE="pln-svc"
+
+  local args
+  args="$(getopt -o dhe:k:w: -l "debug,help,environment,kcp,workspace" -n "$0" -- "$@")"
+  eval set -- "$args"
+  while true; do
+    case "$1" in
+    -e | --environment)
+      shift
+      ENVIRONMENT=$1
+      case $1 in
+      local) ;;
+      *)
+        if [ ! -d "$SCRIPT_DIR/environments/$ENVIRONMENT" ]; then
+          echo "Unsupported environment: $1" >&2
+          echo "" >&2
+          usage
+          exit 1
+        fi
+        ;;
+      esac
+      ;;
+    -k | --kcp)
+      shift
+      case $1 in
+      stable | unstable)
+        KCP_ENV="kcp-$1"
+        ;;
+      *)
+        echo "Unsupported environment: $1" >&2
+        echo "" >&2
+        usage
+        exit 1
+        ;;
+      esac
+      ;;
+    -w | --workspace)
+      shift
+      KCP_WORKSPACE="$1"
+      ;;
+    -d | --debug)
+      set -x
+      DEBUG="-d"
+      ;;
+    -h | --help)
+      usage
+      exit 0
+      ;;
+    --)
+      # End of arguments
+      break
+      ;;
+    *)
+      echo "Unknown argument: $1"
+      usage
+      exit 1
+      ;;
+    esac
+    shift
+  done
+}
+
+argocd_local() {
+    mkdir -p "$ARGOCD_DIR"
+    argocd --config "$ARGOCD_DIR/config" "$@"
+}
+
+kcp_config() {
+  mkdir -p "$KUBECONFIG_DIR"
+  KUBECONFIG="$KUBECONFIG_KCP" "$@"
+}
+
+plnsvc_config() {
+  mkdir -p "$KUBECONFIG_DIR"
+  KUBECONFIG="$KUBECONFIG_PLNSVC" "$@"
+}
+
+init() {
+  local ns_locator
+
+  # Create the required CRs
+  if ! kcp_config kubectl get namespace default >/dev/null 2>&1; then
+    kcp_config kubectl create namespace default
+  fi
+  if ! kcp_config kubectl get serviceaccount default >/dev/null 2>&1; then
+    kcp_config kubectl create serviceaccount default
+  fi
+
+  # Retrieve the KCP namespace id
+  ns_locator="{\"logical-cluster\":\"$(
+    kcp_config kubectl kcp workspace current | cut -d\" -f2
+  )\",\"namespace\":\"default\"}"
+  # Loop is necessary as it takes KCP time to create the namespace
+  while ! plnsvc_config kubectl get ns -o yaml | grep -q "$ns_locator"; do
+    sleep 2
+  done
+  # Maybe yq could be used effectively here
+  KCP_NS_ID="$(plnsvc_config kubectl get ns -l workloads.kcp.dev/cluster=plnsvc -o yaml \
+    | grep -E " kcp.dev/namespace-locator:| name: kcp" \
+    | grep -A 1 "$ns_locator" | tail -1 \
+    | sed 's/.*: kcp/kcp/'
+  )"
+}
```

Comment:
`k get ns -l workloads.kcp.dev/cluster=plnsvc -o json| jq '.items[].metadata | select(.annotations."kcp.dev/namespace-locator" | contains("default")) | .name'`

Using jq because filtering based on name inside a complex string (namespace-locator annotation value) is not supported with native kube.

Reply:
Thanks for providing the command.

Reply:
I see that you're using ns_locator to check if the namespace is existing or not from line 134 where you're creating the default namespace. But, I think we don't have to create default namespace as part of the './test.sh' script. When we run deploy.sh, I think the default namespace would already be created in kcp and thus the corresponding hash generated ns in the physical cluster.

Please verify this and if true, we can completely remove most part of init() func except KCP_NS_ID and service account creation.

Reply:
ckcp and kind are currently broken for me. I suggest we leave it like that for the moment, and remove it once we can test it properly.
The issue seems minor enough to not hold the merge.